File: android/core/src/test/kotlin/me/raddatz/cindy/core/workout/WorkoutStateMachineTests.kt
package me.raddatz.cindy.core.workout

import me.raddatz.cindy.core.Exercise
import me.raddatz.cindy.core.WorkoutPlan
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

class WorkoutStateMachineTests {
    private fun runningMachine(): WorkoutStateMachine {
        val machine = WorkoutStateMachine()
        machine.beginCountdown()
        machine.start()
        machine.activate()
        return machine
    }

    @Test
    fun startsWithPullUpsInTransition() {
        val machine = WorkoutStateMachine()
        machine.beginCountdown()
        assertEquals(WorkoutPhase.COUNTDOWN, machine.phase)
        assertEquals(listOf<WorkoutEvent>(WorkoutEvent.Started), machine.start())
        assertEquals(WorkoutPhase.TRANSITION, machine.phase)
        assertEquals(Exercise.PULL_UP, machine.exercise)
        assertEquals(listOf<WorkoutEvent>(WorkoutEvent.ExerciseStarted(Exercise.PULL_UP)), machine.activate())
        assertEquals(WorkoutPhase.ACTIVE, machine.phase)
    }

    @Test
    fun repsAreIgnoredWhileInTransition() {
        val machine = WorkoutStateMachine()
        machine.beginCountdown()
        machine.start()
        assertTrue(machine.registerRep().isEmpty())
        assertEquals(0, machine.repCount)
    }

    @Test
    fun fullRoundAdvancesExercisesAndRound() {
        val machine = runningMachine()
        repeat(4) { machine.registerRep() }
        assertEquals(4, machine.repCount)
        val fifth = machine.registerRep()
        assertTrue(fifth.contains(WorkoutEvent.ExerciseCompleted(Exercise.PULL_UP, next = Exercise.PUSH_UP)))
        assertEquals(WorkoutPhase.TRANSITION, machine.phase)
        assertEquals(Exercise.PUSH_UP, machine.exercise)
        assertEquals(0, machine.repCount)
        assertEquals(WorkoutScore(rounds = 0, reps = 5), machine.score)

        machine.activate()
        repeat(10) { machine.registerRep() }
        assertEquals(Exercise.SQUAT, machine.exercise)
        assertEquals(15, machine.score.reps)

        machine.activate()
        val events = mutableListOf<WorkoutEvent>()
        repeat(15) { events += machine.registerRep() }
        assertTrue(events.contains(WorkoutEvent.RoundCompleted(1)))
        assertEquals(Exercise.PULL_UP, machine.exercise)
        assertEquals(1, machine.completedRounds)
        assertEquals(2, machine.currentRound)
        assertEquals(WorkoutScore(rounds = 1, reps = 0), machine.score)
    }

    @Test
    fun scoreNotationFollowsCrossFitConvention() {
        assertEquals("14 + 7", WorkoutScore(rounds = 14, reps = 7).notation)
        assertEquals("14", WorkoutScore(rounds = 14, reps = 0).notation)
        assertEquals(77, WorkoutScore(rounds = 2, reps = 17).totalReps)
    }

    @Test
    fun manualPlusOneCompletesExercise() {
        val machine = runningMachine()
        repeat(4) { machine.registerRep() }
        val events = machine.adjust(by = 1)
        assertTrue(events.contains(WorkoutEvent.ExerciseCompleted(Exercise.PULL_UP, next = Exercise.PUSH_UP)))
        assertEquals(WorkoutPhase.TRANSITION, machine.phase)
    }

    @Test
    fun manualMinusOneStepsBackIntoPreviousExercise() {
        val machine = runningMachine()
        repeat(5) { machine.registerRep() } // now push-ups, transition
        val events = machine.adjust(by = -1)
        assertTrue(events.contains(WorkoutEvent.ExerciseReopened(Exercise.PULL_UP)))
        assertEquals(Exercise.PULL_UP, machine.exercise)
        assertEquals(4, machine.repCount)
        assertEquals(WorkoutScore(rounds = 0, reps = 4), machine.score)
    }

    @Test
    fun minusOneAcrossRoundBoundaryReopensRound() {
        val machine = runningMachine()
        repeat(5) { machine.registerRep() }
        machine.activate()
        repeat(10) { machine.registerRep() }
        machine.activate()
        repeat(15) { machine.registerRep() }
        assertEquals(1, machine.completedRounds)
        val events = machine.adjust(by = -1)
        assertTrue(events.contains(WorkoutEvent.RoundReopened(1)))
        assertEquals(0, machine.completedRounds)
        assertEquals(Exercise.SQUAT, machine.exercise)
        assertEquals(14, machine.repCount)
    }

    @Test
    fun minusOneAtVeryStartDoesNothing() {
        val machine = runningMachine()
        assertTrue(machine.adjust(by = -1).isEmpty())
        assertEquals(0, machine.repCount)
        assertEquals(Exercise.PULL_UP, machine.exercise)
    }

    @Test
    fun pauseAndResumeGoThroughTransition() {
        val machine = runningMachine()
        machine.registerRep()
        machine.pause()
        assertEquals(WorkoutPhase.PAUSED, machine.phase)
        assertTrue(machine.registerRep().isEmpty())
        assertTrue(machine.adjust(by = 1).contains(WorkoutEvent.RepCounted(Exercise.PULL_UP, count = 2)))
        machine.resume()
        assertEquals(WorkoutPhase.TRANSITION, machine.phase)
        machine.activate()
        assertEquals(WorkoutPhase.ACTIVE, machine.phase)
        assertEquals(2, machine.repCount)
    }

    @Test
    fun finishStopsCounting() {
        val machine = runningMachine()
        machine.registerRep()
        assertEquals(listOf<WorkoutEvent>(WorkoutEvent.Finished), machine.finish())
        assertEquals(WorkoutPhase.FINISHED, machine.phase)
        assertTrue(machine.registerRep().isEmpty())
        assertTrue(machine.finish().isEmpty())
        assertEquals(WorkoutScore(rounds = 0, reps = 1), machine.score)
    }

    // Changing the plan mid-workout

    /** Pull-ups done, `pushUps` push-ups counted. */
    private fun machineInPushUps(pushUps: Int): WorkoutStateMachine {
        val machine = runningMachine()
        repeat(5) { machine.registerRep() }
        machine.activate()
        repeat(pushUps) { machine.registerRep() }
        return machine
    }

    @Test
    fun changedTargetKeepsTheRepsOfTheCurrentExercise() {
        val machine = machineInPushUps(4)
        val plan = WorkoutPlan.cindy.withTarget(8, Exercise.PUSH_UP)
        assertTrue(machine.replacePlan(plan).isEmpty())
        assertEquals(Exercise.PUSH_UP, machine.exercise)
        assertEquals(4, machine.repCount)
        assertEquals(WorkoutPhase.ACTIVE, machine.phase)
        assertEquals(WorkoutScore(rounds = 0, reps = 9, repsPerRound = 28), machine.score)
    }

    @Test
    fun targetBelowTheRepsDoneFinishesTheExercise() {
        val machine = machineInPushUps(7)
        machine.pause()
        val plan = WorkoutPlan.cindy.withTarget(5, Exercise.PUSH_UP)
        assertEquals(
            listOf<WorkoutEvent>(WorkoutEvent.ExerciseCompleted(Exercise.PUSH_UP, next = Exercise.SQUAT)),
            machine.replacePlan(plan),
        )
        assertEquals(Exercise.SQUAT, machine.exercise)
        assertEquals(0, machine.repCount)
        assertEquals(WorkoutPhase.PAUSED, machine.phase)
    }

    @Test
    fun removingTheCurrentExerciseMovesOnToTheNextOne() {
        val machine = machineInPushUps(3)
        val plan = WorkoutPlan.cindy.withEnabled(Exercise.PUSH_UP, false)
        assertTrue(machine.replacePlan(plan).isEmpty())
        assertEquals(Exercise.SQUAT, machine.exercise)
        assertEquals(0, machine.repCount)
        assertEquals(WorkoutPhase.TRANSITION, machine.phase)
        assertEquals(WorkoutScore(rounds = 0, reps = 5, repsPerRound = 20), machine.score)
    }

    @Test
    fun removingTheRestOfTheRoundCompletesIt() {
        val machine = machineInPushUps(10)
        machine.activate()
        repeat(3) { machine.registerRep() }
        val plan = WorkoutPlan.cindy.withEnabled(Exercise.SQUAT, false)
        assertEquals(listOf<WorkoutEvent>(WorkoutEvent.RoundCompleted(1)), machine.replacePlan(plan))
        assertEquals(Exercise.PULL_UP, machine.exercise)
        assertEquals(2, machine.currentRound)
        assertEquals(WorkoutScore(rounds = 1, reps = 0, repsPerRound = 15), machine.score)
    }

    @Test
    fun addedExerciseJoinsTheCurrentRound() {
        val machine = machineInPushUps(10)
        val plan = WorkoutPlan.cindy.withEnabled(Exercise.PLANK, true)
        machine.replacePlan(plan)
        machine.activate()
        // The plank waits for the end of the AMRAP: the squats still complete the round.
        val events = (0 until 15).flatMap { machine.registerRep() }
        assertTrue(events.contains(WorkoutEvent.RoundCompleted(1)))
        assertEquals(Exercise.PULL_UP, machine.exercise)
    }

    @Test
    fun thePlankOnlyFollowsARunningWorkoutWithAPlank() {
        val machine = machineInPushUps(0)
        machine.beginPlank()
        assertNotEquals(WorkoutPhase.PLANK, machine.phase) // Cindy has no plank
        machine.replacePlan(WorkoutPlan.cindy.withEnabled(Exercise.PLANK, true))
        machine.beginPlank()
        assertEquals(WorkoutPhase.PLANK, machine.phase)
        assertFalse(machine.isRunning)
    }

    @Test
    fun finishedWorkoutKeepsItsPlan() {
        val machine = runningMachine()
        machine.finish()
        assertTrue(machine.replacePlan(WorkoutPlan.withoutPullUps).isEmpty())
        assertEquals(WorkoutPlan.cindy, machine.plan)
    }
}

File: ios/CindyTests/WorkoutStateMachineTests.swift
import Testing
@testable import Cindy

struct WorkoutStateMachineTests {
    private func runningMachine() -> WorkoutStateMachine {
        let machine = WorkoutStateMachine()
        machine.beginCountdown()
        machine.start()
        machine.activate()
        return machine
    }

    @Test func startsWithPullUpsInTransition() {
        let machine = WorkoutStateMachine()
        machine.beginCountdown()
        #expect(machine.phase == .countdown)
        #expect(machine.start() == [.started])
        #expect(machine.phase == .transition)
        #expect(machine.exercise == .pullUp)
        #expect(machine.activate() == [.exerciseStarted(.pullUp)])
        #expect(machine.phase == .active)
    }

    @Test func repsAreIgnoredWhileInTransition() {
        let machine = WorkoutStateMachine()
        machine.beginCountdown()
        machine.start()
        #expect(machine.registerRep().isEmpty)
        #expect(machine.repCount == 0)
    }

    @Test func fullRoundAdvancesExercisesAndRound() {
        let machine = runningMachine()
        for _ in 0..<4 { machine.registerRep() }
        #expect(machine.repCount == 4)
        let fifth = machine.registerRep()
        #expect(fifth.contains(.exerciseCompleted(.pullUp, next: .pushUp)))
        #expect(machine.phase == .transition)
        #expect(machine.exercise == .pushUp)
        #expect(machine.repCount == 0)
        #expect(machine.score == WorkoutScore(rounds: 0, reps: 5))

        machine.activate()
        for _ in 0..<10 { machine.registerRep() }
        #expect(machine.exercise == .squat)
        #expect(machine.score.reps == 15)

        machine.activate()
        var events: [WorkoutEvent] = []
        for _ in 0..<15 { events += machine.registerRep() }
        #expect(events.contains(.roundCompleted(1)))
        #expect(machine.exercise == .pullUp)
        #expect(machine.completedRounds == 1)
        #expect(machine.currentRound == 2)
        #expect(machine.score == WorkoutScore(rounds: 1, reps: 0))
    }

    @Test func scoreNotationFollowsCrossFitConvention() {
        #expect(WorkoutScore(rounds: 14, reps: 7).notation == "14 + 7")
        #expect(WorkoutScore(rounds: 14, reps: 0).notation == "14")
        #expect(WorkoutScore(rounds: 2, reps: 17).totalReps == 77)
    }

    @Test func manualPlusOneCompletesExercise() {
        let machine = runningMachine()
        for _ in 0..<4 { machine.registerRep() }
        let events = machine.adjust(by: 1)
        #expect(events.contains(.exerciseCompleted(.pullUp, next: .pushUp)))
        #expect(machine.phase == .transition)
    }

    @Test func manualMinusOneStepsBackIntoPreviousExercise() {
        let machine = runningMachine()
        for _ in 0..<5 { machine.registerRep() } // now push-ups, transition
        let events = machine.adjust(by: -1)
        #expect(events.contains(.exerciseReopened(.pullUp)))
        #expect(machine.exercise == .pullUp)
        #expect(machine.repCount == 4)
        #expect(machine.score == WorkoutScore(rounds: 0, reps: 4))
    }

    @Test func minusOneAcrossRoundBoundaryReopensRound() {
        let machine = runningMachine()
        for _ in 0..<5 { machine.registerRep() }
        machine.activate()
        for _ in 0..<10 { machine.registerRep() }
        machine.activate()
        for _ in 0..<15 { machine.registerRep() }
        #expect(machine.completedRounds == 1)
        let events = machine.adjust(by: -1)
        #expect(events.contains(.roundReopened(1)))
        #expect(machine.completedRounds == 0)
        #expect(machine.exercise == .squat)
        #expect(machine.repCount == 14)
    }

    @Test func minusOneAtVeryStartDoesNothing() {
        let machine = runningMachine()
        #expect(machine.adjust(by: -1).isEmpty)
        #expect(machine.repCount == 0)
        #expect(machine.exercise == .pullUp)
    }

    @Test func pauseAndResumeGoThroughTransition() {
        let machine = runningMachine()
        machine.registerRep()
        machine.pause()
        #expect(machine.phase == .paused)
        #expect(machine.registerRep().isEmpty)
        #expect(machine.adjust(by: 1).contains(.repCounted(.pullUp, count: 2)))
        machine.resume()
        #expect(machine.phase == .transition)
        machine.activate()
        #expect(machine.phase == .active)
        #expect(machine.repCount == 2)
    }

    @Test func finishStopsCounting() {
        let machine = runningMachine()
        machine.registerRep()
        #expect(machine.finish() == [.finished])
        #expect(machine.phase == .finished)
        #expect(machine.registerRep().isEmpty)
        #expect(machine.finish().isEmpty)
        #expect(machine.score == WorkoutScore(rounds: 0, reps: 1))
    }

    // MARK: - Changing the plan mid-workout

    /// Pull-ups done, `pushUps` push-ups counted.
    private func machineInPushUps(_ pushUps: Int) -> WorkoutStateMachine {
        let machine = runningMachine()
        for _ in 0..<5 { machine.registerRep() }
        machine.activate()
        for _ in 0..<pushUps { machine.registerRep() }
        return machine
    }

    @Test func changedTargetKeepsTheRepsOfTheCurrentExercise() {
        let machine = machineInPushUps(4)
        var plan = WorkoutPlan.cindy
        plan.setTarget(8, for: .pushUp)
        #expect(machine.replacePlan(plan).isEmpty)
        #expect(machine.exercise == .pushUp)
        #expect(machine.repCount == 4)
        #expect(machine.phase == .active)
        #expect(machine.score == WorkoutScore(rounds: 0, reps: 9, repsPerRound: 28))
    }

    @Test func targetBelowTheRepsDoneFinishesTheExercise() {
        let machine = machineInPushUps(7)
        machine.pause()
        var plan = WorkoutPlan.cindy
        plan.setTarget(5, for: .pushUp)
        #expect(machine.replacePlan(plan) == [.exerciseCompleted(.pushUp, next: .squat)])
        #expect(machine.exercise == .squat)
        #expect(machine.repCount == 0)
        #expect(machine.phase == .paused)
    }

    @Test func removingTheCurrentExerciseMovesOnToTheNextOne() {
        let machine = machineInPushUps(3)
        var plan = WorkoutPlan.cindy
        plan.setEnabled(.pushUp, false)
        #expect(machine.replacePlan(plan).isEmpty)
        #expect(machine.exercise == .squat)
        #expect(machine.repCount == 0)
        #expect(machine.phase == .transition)
        #expect(machine.score == WorkoutScore(rounds: 0, reps: 5, repsPerRound: 20))
    }

    @Test func removingTheRestOfTheRoundCompletesIt() {
        let machine = machineInPushUps(10)
        machine.activate()
        for _ in 0..<3 { machine.registerRep() }
        var plan = WorkoutPlan.cindy
        plan.setEnabled(.squat, false)
        #expect(machine.replacePlan(plan) == [.roundCompleted(1)])
        #expect(machine.exercise == .pullUp)
        #expect(machine.currentRound == 2)
        #expect(machine.score == WorkoutScore(rounds: 1, reps: 0, repsPerRound: 15))
    }

    @Test func addedExerciseJoinsTheCurrentRound() {
        let machine = machineInPushUps(10)
        var plan = WorkoutPlan.cindy
        plan.setEnabled(.plank, true)
        machine.replacePlan(plan)
        machine.activate()
        // The plank waits for the end of the AMRAP: the squats still complete the round.
        let events = (0..<15).flatMap { _ in machine.registerRep() }
        #expect(events.contains(.roundCompleted(1)))
        #expect(machine.exercise == .pullUp)
    }

    @Test func thePlankOnlyFollowsARunningWorkoutWithAPlank() {
        let machine = machineInPushUps(0)
        machine.beginPlank()
        #expect(machine.phase != .plank) // Cindy has no plank
        var plan = WorkoutPlan.cindy
        plan.setEnabled(.plank, true)
        machine.replacePlan(plan)
        machine.beginPlank()
        #expect(machine.phase == .plank)
        #expect(machine.isRunning == false)
    }

    @Test func finishedWorkoutKeepsItsPlan() {
        let machine = runningMachine()
        machine.finish()
        #expect(machine.replacePlan(.withoutPullUps).isEmpty)
        #expect(machine.plan == .cindy)
    }
}
